x-pack/solutions/observability/plugins/apm/public/components/shared/trace_waterfall/use_trace_waterfall.ts

```diff
@@ -89,18 +99,34 @@ export function createColorLookupMap(legends: IWaterfallLegend[]): Map<string, s
   return new Map(legends.map((legend) => [`${legend.type}:${legend.value}`, legend.color]));
 }
 
-export function getTraceParentChildrenMap(traceItems: TraceItem[]) {
-  const traceMap = traceItems.reduce<Record<string, TraceItem[]>>((acc, item) => {
-    if (!item.parentId) {
-      acc.root = [item];
+export function getTraceParentChildrenMap(traceItems: TraceItem[], filteredTrace: boolean) {
+  if (traceItems.length === 0) {
+    return {};
+  }
+
+  const traceMap: Record<string, TraceItem[]> = {};
+
+  traceItems.reduce<Record<string, TraceItem[]>>((acc, item) => {
+    if (item.parentId) {
+      (acc[item.parentId] ??= []).push(item);
     } else {
-      if (!acc[item.parentId]) {
-        acc[item.parentId] = [];
-      }
-      acc[item.parentId].push(item);
+      (acc.root ??= [])[0] = item;
     }
     return acc;
-  }, {});
+  }, traceMap);
+
+  // If filtered trace, elect the earliest arriving span as the root if there is no root found already
+  if (filteredTrace && !traceMap.root) {
+    const root = traceItems
+      .slice(1)
+      .reduce(
+        (acc, span) =>
+          acc.timestampUs <= span.timestampUs || acc.id === span.parentId ? acc : span,
+        traceItems[0]
+      );
+
+    traceMap.root = [root];
+  }
```

Comment:
Should we move this logic to the server too? Maybe we could add a new property called `isRoot` to the server response payload and the logic will be:

When it's not filtered: isRoot is true when parent.id is undefined
When it's filtered: isRoot is true following the logic done here.

This way the UI will never be responsible to nominee a root, it's always a server responsability.